File: dotnet/Barchart.BinarySerializer/Entities/Keys/EntityKey.cs
namespace Barchart.BinarySerializer.Entities.Keys;

/// <summary>
///     Represents a unique key for an entity used in (de)serialization process.
/// </summary>
/// <typeparam name="TEntity">
///     The type of the entity.
/// </typeparam>
public class EntityKey<TEntity> : IEntityKey<TEntity>, IEquatable<EntityKey<TEntity>> where TEntity : class, new()
{
    #region Fields
    
    private readonly object _key;
    
    #endregion
    
    #region Construtor(s)
    
    /// <summary>
    ///     Initializes a new instance of the <see cref="EntityKey{TEntity}"/> class.
    /// </summary>
    /// <param name="key">
    ///     The key value.
    /// </param>
    /// <exception cref="ArgumentNullException">
    ///     Thrown if the specified key is <see langword="null"/>.
    /// </exception>
    public EntityKey(object key)
    {
        ArgumentNullException.ThrowIfNull(key, nameof(key));
        
        _key = key;
    }
    
    #endregion
    
    #region Methods

    /// <inheritdoc />
    public bool Equals(EntityKey<TEntity>? other)
    {
        if (other == null)
        {
            return false;
        }
        
        return _key.Equals(other._key);
    }

    /// <inheritdoc />
    public override bool Equals(object? obj)
    {
        if (obj is not EntityKey<TEntity> other)
        {
            return false;
        }
        
        return Equals(other);
    }
    
    /// <inheritdoc />
    public override int GetHashCode()
    {
        return _key.GetHashCode();
    }

    /// <inheritdoc />
    public override string ToString()
    {
        return $"{base.ToString()}, (key={_key})";
    }

    #endregion
}

File: typescript/src/entities/keys/entity-key.ts
import Day from "@barchart/common-js/lang/Day";
import Enum from "@barchart/common-js/lang/Enum";
import * as is from "@barchart/common-js/lang/is";
import Big from "big.js";

import { EntityKeyDefinition } from "./entity-key.interface";
import { ArgumentNullException } from "../../exceptions/argument-null-exception";

/**
 * Represents a unique key for an entity used in (de)serialization process.
 *
 * @public
 * @exported
 * @template TEntity - The type of the entity.
 * @implements EntityKeyDefinition<TEntity>
 * @param {object} key - The unique key of the entity.
 */
export class EntityKey<TEntity extends object> implements EntityKeyDefinition<TEntity> {
    private readonly _key: object;

    constructor(key: object) {
        if (is.nil(key) || is.undef(key)) {
            throw new ArgumentNullException("key");
        }

        this._key = key;
    }

    equals(other: EntityKeyDefinition<TEntity> | null): boolean {
        if (!(other instanceof EntityKey)) {
            return false;
        }

        return EntityKey.equalsValue(this._key, other._key);
    }

    toString(): string {
        return `${this.constructor.name}, (key=${this._key})`;
    }

    private static equalsValue(left: unknown, right: unknown): boolean {
        if (Array.isArray(left) && Array.isArray(right)) {
            return left.length === right.length && left.every((value, index) => EntityKey.equalsValue(value, right[index]));
        }

        return left === right || (is.nan(left) && is.nan(right)) || (left instanceof Date && right instanceof Date && left.getTime() === right.getTime()) || (left instanceof Day && right instanceof Day && left.getIsEqual(right)) || (left instanceof Big && right instanceof Big && left.eq(right)) || (left instanceof Enum && right instanceof Enum && left.equals(right));
    }
}
